[Data] Move extension types to `ray.data`

zzchun opened the pull request:
## Description

Relocate extension types from `ray.air.util` to `ray.data._internal` to reduce module dependencies. This includes moving `object_extensions` and `tensor_extensions`, updating imports codebase-wide, fixing import errors and test failures, and planning further cleanups to fully decouple `ray.data` from `ray.air`

The changes mainly included: 

`ray.air.util.tensor_extensions` -> `ray.data._internal.tensor_extensions`
`ray.air.util.object_extensions` -> `ray.data._internal.object_extensions`
`ray.air._internal.torch_utils`  -> `ray.data.util.torch_utils`
`python/ray/air/util/transform_pyarrow.py`  ->  `ray/data/_internal/utils/transform_pyarrow.py`
`ray.air.util -> ray.data.util`

## Related issues
Closes #59418

bveeramani commented on python/ray/data/_internal/tensor_extensions/pandas.py:
```diff
@@ -1018,25 +1026,21 @@ def take(
+        ...     # the data, before passing to take.
+        ...     result = take(data, indices, fill_value=fill_value,
+        ...                   allow_fill=allow_fill)
+        ...     return self._from_sequence(result, dtype=self.dtype)  # doctest: +SKIP
```
Since this is a long code snippet, I think this is easier to read as a `.. testcode` or `.. code-block`. Could you revert? This is also unrelated to the PR, and might complicate the git blame

bveeramani reviewed (commented):
Ty! 

Overall LGTM! Just a couple minor comments about the docstrings

richardliaw commented on python/ray/air/_internal/tensorflow_utils.py:
```diff
@@ -5,7 +5,7 @@
 
 from ray.air.util.data_batch_conversion import _unwrap_ndarray_object_type_if_needed
-from ray.air.util.tensor_extensions.arrow import get_arrow_extension_tensor_types
+from ray.data._internal.tensor_extensions.arrow import get_arrow_extension_tensor_types
```
I suspect it's fairly manageable... ray air also has various train things, we just need to move the data stuff

zzchun committed:
[Data]Move extension types to ray.data

ray.air.util.objection_extensions -> ray.data._internal.object_extensions
ray.air.util.tensor_extensions -> ray.data._internal.tensor_extensions

Signed-off-by: will <zzchun8@gmail.com>

zzchun commented:
/cc @richardliaw @bveeramani 

👋 Could you please take a look at this PR when you have time?

The PR has been facing frequent merge conflicts with recently merged changes

Thank you!

bveeramani commented on python/ray/air/_internal/tensorflow_utils.py:
```diff
@@ -1,14 +1,9 @@
-    from ray.data._internal.pandas_block import PandasBlockSchema
 
 
 def convert_ndarray_to_tf_tensor(
```
This function is now duplicated across `ray.air` and `ray.data`, but its not dead code. 

In a follow-up, I think we should remove this and update the reference in `tensorflow_predictor` to use the one defined in `ray.data`.

bveeramani commented on python/ray/data/util/torch_utils.py:
```diff
@@ -0,0 +1,618 @@
+    return result
+
+
+def convert_ndarray_batch_to_torch_tensor_batch(
```
I don't think we use this anywhere in Ray Data. I think we should remove this in a follow-up

bveeramani commented on python/ray/data/util/torch_utils.py:
```diff
@@ -0,0 +1,618 @@
+    return batch
+
+
+def load_torch_model(
```
Ditto

bveeramani reviewed (approved):
LGTM! ty for the contribution!

bveeramani commented:
@zzchun would you mind updating the PR description with the latest changes?

matthewdeng commented on python/ray/air/util/data_batch_conversion.py:
```diff
@@ -217,10 +217,10 @@ def _convert_batch_type_to_numpy(
         return data
     elif pyarrow is not None and isinstance(data, pyarrow.Table):
-        from ray.air.util.tensor_extensions.arrow import (
+        from ray.data._internal.arrow_ops import transform_pyarrow
```
Why is this still importing from data? Should this util also be moved to data?

richardliaw replied:
oh, the reason this is here is because this utility is still dependent from ray.train.predictors, so this is a duplicate. How about let's just address this in a later PR when we remove predictors?

bveeramani commented:
@zzchun I merged the PR to avoid merge conflicts, but pls update the PR description when you get a chance so it's representative of the actual diff!
